blocks/datasets/__init__.py

```diff
@@ -0,0 +1,223 @@
+
+        Notes
+        -----
+        This is implemented because simply repeatedly calling :meth:`open`
+        could result in file handlers not being closed, and to increase
+        performance by performing a seek-operation instead of re-opening
+        the file.
+
+        """
+        pass
+
+    @abstractmethod
+    def get_data(self, state=None, request=None, sources=None):
+        """Request data from the dataset.
+
+        Parameters
+        ----------
+        state : object, optional
+            The state as returned by the :meth:`open` method. The dataset
+            can use this to e.g. interact with files when needed.
+        request : object, optional
+            If supported, the request for a particular part of the data
+            e.g. the number of examples to return, or the indices of a
+            particular minimbatch of examples.
+        sources : tuple of strings, optional
+            The data sources to return. By default all data sources are
+            returned.
+
+        .. todo::
+
+           A way for the dataset to communicate which kind of requests it
+           accepts, and a way to communicate what kind of request is being
+           sent when supporting multiple.
+
+        Returns
+        -------
+        tuple
+            A tuple of data.
+
+        """
+        raise NotImplementedError
+
+
+class DataStream(object):
+    """A stream of data.
+
+    A data stream implements the ``__iter__`` protocol, returning a new
+    iterator over the data each time it gets called (in the case of finite
+    data). A single pass over one of these iterators represents an *epoch*.
+
+    The data stream also opens a given dataset and maintains the state
+    (e.g. file handles) if necessary.
+
+    A data stream can provide an iterator over a dataset, but it can also
+    wrap another datastream e.g. to provide caching or some sort of
+    pre-processing.
+
+    Parameters
+    ----------
+    data : :class:`Dataset` or :class:`DataStream`
+        An object that implements the :meth:`open`, :meth:`reset` and
+        :meth:`get_data` methods. Usually this is a :class:`Dataset`
+        instance, but when chaining iterators to e.g. perform caching this
+        can also be a :class:`DataStream` object.
+
+    """
+    def __init__(self, data, iteration_scheme=None, sources=None):
+        data_state = data.open()
+        update_instance(self, locals())
+
+    def __iter__(self):
+        self.data_state = self.data.reset(self.data_state)
+        return DataIterator(self,
+                            iter(self.iteration_scheme)
+                            if self.iteration_scheme else None, self.sources)
+
+    @property
+    def sources(self):
+        if getattr(self, '_sources', None) is None:
+            return self.data.sources
+
+    @sources.setter
+    def sources(self, sources):
+        self._sources = sources
+
+    def open(self):
```

Comment:
I agree with Jan that what the interface data streams provides to the 
rest of the library should be isolated in a base class. I had the same 
confusions as he had.

B.t.w., the interface you propose looks nice when `for batch in stream` 
iterates over the same data every time called, but what if an epoch is 
just a part of the data, not the whole data? Can `__iter__` method of a 
data stream return iterators over different data?

On 09.01.2015 11:01, Bart van Merriënboer wrote:


> In blocks/datasets/__init__.py 
> <https://github.com/bartvm/blocks/pull/91#discussion-diff-22704830>:
>
> > +    def __iter__(self):
> > +        self.data_state = self.data.reset(self.data_state)
> > +        return DataIterator(self,
> > +                            iter(self.iteration_scheme)
> > +                            if self.iteration_scheme else None, self.sources)
> > +
> > +    @property
> > +    def sources(self):
> > +        if getattr(self, '_sources', None) is None:
> > +            return self.data.sources
> > +
> > +    @sources.setter
> > +    def sources(self, sources):
> > +        self._sources = sources
> > +
> > +    def open(self):
> Documentation and more implementations will make that clear. On Jan 9, 
> 2015 8:11 AM, "Jan Chorowski" <notifications@github.com> wrote:
> … <#>
> In blocks/datasets/__init__.py 
> <https://github.com/bartvm/blocks/pull/91#discussion-diff-22703899>: > 
> + def __iter__(self): > + self.data_state = 
> self.data.reset(self.data_state) > + return DataIterator(self, > + 
> iter(self.iteration_scheme) > + if self.iteration_scheme else None, 
> self.sources) > + > + @property <https://github.com/property> > + def 
> sources(self): > + if getattr(self, '_sources', None) is None: > + 
> return self.data.sources > + > + @sources 
> <https://github.com/sources>.setter > + def sources(self, sources): > 
> + self._sources = sources > + > + def open(self): Yes. But it will be 
> more visible to the (human) users that the rest of the library doesn't 
> depend on open, reset and other methods. If you make them part of the 
> base interface, then you don't know whether you need to implement 
> them, or not. … <#14acd88dbbdc4b1b_> On 1/8/2015 11:05 PM, Bart van 
> Merriënboer wrote: In blocks/datasets/__init__.py < 
> https://github.com/bartvm/blocks/pull/91#discussion-diff-22686619>: > 
> + def __iter__(self): > + self.data_state = 
> self.data.reset(self.data_state) > + return DataIterator(self, > + 
> iter(self.iteration_scheme) > + if self.iteration_scheme else None, 
> self.sources) > + > + @property <https://github.com/property> 
> <https://github.com/property> > + def sources(self): > + if 
> getattr(self, '_sources', None) is None: > + return self.data.sources 
> > + > + @sources <https://github.com/sources> 
> <https://github.com/sources>.setter 
> <https://github.com/sources%3E.setter> > + def sources(self, sources): 
> > + self._sources = sources > + > + def open(self): But |open| is 
> required even on the base class because it's used when the data stream 
> is wrapped by another data stream. The parent data stream will call 
> |open| on the wrapped data stream, which will propagate down all the 
> data streams till |open| is called on the dataset. — Reply to this 
> email directly or view it on GitHub < 
> https://github.com/bartvm/blocks/pull/91/files#r22686619>. — Reply to 
> this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/91/files#r22703899>.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/91/files#r22704830>.
>